File: algorithms/trees/kd-tree/java/KDTree.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class KDTree {
    private Node root;

    public KDTree(List<int[]> points) {
        root = buildTree(points, 0);
    }

    private Node buildTree(List<int[]> points, int axis) {
        if (points.isEmpty()) {
            return null;
        }

        points.sort(Comparator.comparingInt(point -> point[axis]));

        int medianIdx = points.size() / 2;
        int[] median = points.get(medianIdx);

        int nextAxis = (axis + 1) % median.length;
        Node node = new Node(median, axis);
        node.setLeft(buildTree(points.subList(0, medianIdx), nextAxis));
        node.setRight(buildTree(points.subList(medianIdx + 1, points.size()), nextAxis));

        return node;
    }

    public int[] nearestNeighbor(int[] target) {
        int[] best = new int[]{Integer.MAX_VALUE, -1};
        nearestNeighbor(root, target, best);
        return Arrays.copyOf(best, best.length);
    }

    private void nearestNeighbor(Node node, int[] target, int[] best) {
        if (node == null) {
            return;
        }

        int distance = calculateDistance(node.getPoint(), target);
        if (distance < best[0]) {
            best[0] = distance;
            best[1] = node.getPoint()[0];
        }

        int axis = node.getAxis();
        if (target[axis] < node.getPoint()[axis]) {
            nearestNeighbor(node.getLeft(), target, best);
            if ((node.getPoint()[axis] - target[axis]) * (node.getPoint()[axis] - target[axis]) < best[0]) {
                nearestNeighbor(node.getRight(), target, best);
            }
        } else {
            nearestNeighbor(node.getRight(), target, best);
            if ((target[axis] - node.getPoint()[axis]) * (target[axis] - node.getPoint()[axis]) < best[0]) {
                nearestNeighbor(node.getLeft(), target, best);
            }
        }
    }

    private int calculateDistance(int[] point1, int[] point2) {
        int distance = 0;
        for (int i = 0; i < point1.length; i++) {
            distance += (point1[i] - point2[i]) * (point1[i] - point2[i]);
        }
        return distance;
    }
}

File: algorithms/trees/kd-tree/cpp/KDTree.cpp
#include "KDTree.h"
#include <algorithm>
#include <limits>

KDTree::KDTree(std::vector<std::vector<int>>& points) {
    root = buildTree(points, 0);
}

Node* KDTree::buildTree(std::vector<std::vector<int>>& points, int axis) {
    if (points.empty()) {
        return nullptr;
    }

    std::sort(points.begin(), points.end(), [axis](const auto& a, const auto& b) {
        return a[axis] < b[axis];
    });

    int medianIdx = points.size() / 2;
    std::vector<int> median = points[medianIdx];

    int nextAxis = (axis + 1) % median.size();
    Node* node = new Node(median, axis);
    node->setLeft(buildTree(std::vector<std::vector<int>>(points.begin(), points.begin() + medianIdx), nextAxis));
    node->setRight(buildTree(std::vector<std::vector<int>>(points.begin() + medianIdx + 1, points.end()), nextAxis));

    return node;
}

std::vector<int> KDTree::nearestNeighbor(std::vector<int>& target) {
    std::vector<int> best = {std::numeric_limits<int>::max(), -1};
    nearestNeighbor(root, target, best);
    return best;
}

void KDTree::nearestNeighbor(Node* node, std::vector<int>& target, std::vector<int>& best) {
    if (node == nullptr) {
        return;
    }

    int distance = calculateDistance(node->getPoint(), target);
    if (distance < best[0]) {
        best[0] = distance;
        best[1] = node->getPoint()[0];
    }

    int axis = node->getAxis();
    if (target[axis] < node->getPoint()[axis]) {
        nearestNeighbor(node->getLeft(), target, best);
        if ((node->getPoint()[axis] - target[axis]) * (node->getPoint()[axis] - target[axis]) < best[0]) {
            nearestNeighbor(node->getRight(), target, best);
        }
    } else {
        nearestNeighbor(node->getRight(), target, best);
        if ((target[axis] - node->getPoint()[axis]) * (target[axis] - node->getPoint()[axis]) < best[0]) {
            nearestNeighbor(node->getLeft(), target, best);
        }
    }
}

int KDTree::calculateDistance(std::vector<int>& point1, std::vector<int>& point2) {
    int distance = 0;
    for (size_t i = 0; i < point1.size(); i++) {
        distance += (point1[i] - point2[i]) * (point1[i] - point2[i]);
    }
    return distance;
}
